src/integrations/blocks/structured-data-blocks.php

```diff
@@ -56,4 +156,197 @@ public function enqueue_block_editor_assets() {
 		$this->asset_manager->enqueue_script( 'structured-data-blocks' );
 		$this->asset_manager->enqueue_style( 'structured-data-blocks' );
 	}
+
+	/**
+	 * Optimizes images in the FAQ blocks.
+	 *
+	 * @param array  $attributes The attributes.
+	 * @param string $content    The content.
+	 *
+	 * @return string The content with images optimized.
+	 */
+	public function optimize_faq_images( $attributes, $content ) {
+		if ( ! isset( $attributes['questions'] ) ) {
+			return $content;
+		}
+
+		return $this->optimize_images( $attributes['questions'], 'answer', $content );
+	}
+
+	/**
+	 * Optimizes images in the How-To blocks.
+	 *
+	 * @param array  $attributes The attributes.
+	 * @param string $content    The content.
+	 *
+	 * @return string The content with images optimized.
+	 */
+	public function optimize_how_to_images( $attributes, $content ) {
+		if ( ! isset( $attributes['steps'] ) ) {
+			return $content;
+		}
+
+		return $this->optimize_images( $attributes['steps'], 'text', $content );
+	}
+
+	/**
+	 * Optimizes images in structured data blocks.
+	 *
+	 * @param array  $elements The list of elements from the block attributes.
+	 * @param string $key      The key in the data to iterate over.
+	 * @param string $content  The content.
+	 *
+	 * @return string The content with images optimized.
+	 */
+	private function optimize_images( $elements, $key, $content ) {
+		global $post;
+
+		$this->add_images_from_attributes_to_used_cache( $post->ID, $elements, $key );
+
+		// Then replace all images with optimized versions in the content.
+		$content = \preg_replace_callback(
+			'/<img[^>]+>/',
+			function ( $matches ) {
+				\preg_match( '/src="([^"]+)"/', $matches[0], $src_matches );
+				if ( ! $src_matches || ! isset( $src_matches[1] ) ) {
+					return $matches[0];
+				}
+				$attachment_id = $this->attachment_src_to_id( $src_matches[1] );
+				if ( $attachment_id === 0 ) {
+					return $matches[0];
+				}
+				$image_size = 'full';
+				\preg_match( '/style="[^"]*width:\s*(\d+)px[^"]*"/', $matches[0], $style_matches );
+				if ( $style_matches && isset( $style_matches[1] ) ) {
+					$width        = (int) $style_matches[1];
+					$meta_data    = \wp_get_attachment_metadata( $attachment_id );
+					$aspect_ratio = ( $meta_data['height'] / $meta_data['width'] );
+					$height       = ( $width * $aspect_ratio );
+					$image_size   = [ $width, $height ];
+				}
```

Comment:
To prevent errors on incomplete data or division by 0 errors
```suggestion
				if ( $style_matches && isset( $style_matches[1] ) ) {
					$width        = (int) $style_matches[1];
					$meta_data    = \wp_get_attachment_metadata( $attachment_id );
					if ( isset( $meta_data['height' ) && isset( $meta_data['width' ) && $meta_data['height'] > 0 && $meta_data['width'] > 0 ) {
						$aspect_ratio = ( $meta_data['height'] / $meta_data['width'] );
						$height       = ( $width * $aspect_ratio );
						$image_size   = [ $width, $height ];
					}
				}
```